components/gitpod-db/src/typeorm/project-db-impl.ts

```diff
@@ -24,13 +24,26 @@ export class ProjectDBImpl implements ProjectDB {
         return (await this.getEntityManager()).getRepository<DBProject>(DBProject);
     }
 
+    async findProjectByCloneUrl(cloneUrl: string): Promise<Project | undefined> {
+        const repo = await this.getRepo();
+        return repo.findOne({ cloneUrl });
+    }
+
+    async findProjectByInstallationId(appInstallationId: string): Promise<Project | undefined> {
+        const repo = await this.getRepo();
+        return repo.findOne({ appInstallationId });
+    }
+
     public async findProjectsByTeam(teamId: string): Promise<Project[]> {
         const repo = await this.getRepo();
         return repo.find({ teamId });
     }
 
     public async createProject(name: string, cloneUrl: string, teamId: string, appInstallationId: string): Promise<Project> {
         const repo = await this.getRepo();
+        if (repo.findOne({ cloneUrl })) {
+            throw new Error('A project with the same clone URL already exists');
+        }
 
         const project: Project = {
             id: uuidv4(),
```

Comment:
Personally I like to decouple such creation logic from the DB, where it is hard to test etc. One example is to move this into a method `Project.new` in `project-protocol.ts` which takes a `Without<Project, 'id' | 'creationTime'>`.

Reply:
you're talking about externalizing the creation of `Project` object to be stored afterwards here. 
then we can if `id` included and add the check. 
that sounds good.